Upstreams changes from Gobra

ArquintL opened the pull request:
Most changes in this PR were made by @Felalolf in the context of Gobra (see [this changeset](https://github.com/viperproject/gobra/compare/802e1cd0c41660cac2a9e3558a4f0de89ff1fb57...5c85dc6030d230136e97a049589169e53068cedb))

jcp19 commented:
It seems that we will soon run into problems with the silver versions once we merge this (the version of the silver submodule might differ from those of the projects that import this plugin). Maybe we should add the update-submodules action here

ArquintL commented:
> It seems that we will soon run into problems with the silver versions once we merge this (the version of the silver submodule might differ from those of the projects that import this plugin). Maybe we should add the update-submodules action here

I completely agree and I've just added this functionality to #2

marcoeilers commented:
So my impression is that this PR contains
- an SBT version update
- at least one bug fix (in ``isUnary``), which previously did not take into account relational predicates
- some (generally good) renaming
- comments explaining what's going on, which is great
- other comments expressing various things, e.g. trying to figure out what the code does (``// TODO REM: maybe this has to be added only so that getUnusedName does not return the name of a declaration``) and some things that I guess Felix wanted to do at some point. 

Am I overlooking anything?
I think the first four are clearly good, I'm not so sure about the last one. I think having a look at some of those and keeping the useful ones might be a good idea.

jcp19 commented:
Is there an ETA for this? I would like to merge it so that we can merge the hyperGobra branch in Gobra so that I can finally fix https://github.com/viperproject/gobra/issues/846, which is the biggest limitation in expressiveness we have in Gobra atm

ArquintL commented:
Sorry for the long wait. I started a while ago and dropped it then for time reasons. I addressed some but not all TODOs. I must admit that some are quite hard to address and/or unclear what they even mean. Without having much clue of what's happening it's not clear to me why certain things are indeed bugs as pointed out by some TODOs.
In any case, I focus on the generation of predicates and heap-dependent functions for relational predicates and tried to clean up their generation, in particular with respect to generating their names by reducing dependencies between the initial pass registering names and the generation phase by using deterministic names (that are guaranteed to be disjoint from names potentially used by a user)

marcoeilers commented:
Looks to me like it didn't break anything.

marcoeilers reviewed (approved):
Okay, let's merge this then.
